ring_theory/multiplicity.lean

```diff
@@ -0,0 +1,325 @@
+⟨λ h, ⟨finite_of_finite_mul_right h, finite_of_finite_mul_left h⟩,
+  λ h, finite_mul hp h.1 h.2⟩
+
+lemma finite_pow {p a : α} (hp : prime p) : Π {k : ℕ} (ha : finite p a), finite p (a ^ k)
+| 0     ha := ⟨0, by simp [mt is_unit_iff_dvd_one.2 hp.2.1]⟩
+| (k+1) ha := by rw [_root_.pow_succ]; exact finite_mul hp ha (finite_pow ha)
+
+protected lemma mul' {p a b : α} (hp : prime p)
+  (h : (multiplicity p (a * b)).dom) :
+  get (multiplicity p (a * b)) h =
+  get (multiplicity p a) ((finite_mul_iff hp).1 h).1 +
+  get (multiplicity p b) ((finite_mul_iff hp).1 h).2 :=
+have hdiva : p ^ get (multiplicity p a) ((finite_mul_iff hp).1 h).1 ∣ a, from spec _,
+have hdivb : p ^ get (multiplicity p b) ((finite_mul_iff hp).1 h).2 ∣ b, from spec _,
+have hpoweq : p ^ (get (multiplicity p a) ((finite_mul_iff hp).1 h).1 +
+    get (multiplicity p b) ((finite_mul_iff hp).1 h).2) =
+    p ^ get (multiplicity p a) ((finite_mul_iff hp).1 h).1 *
+    p ^ get (multiplicity p b) ((finite_mul_iff hp).1 h).2,
+  by simp [_root_.pow_add],
+have hdiv : p ^ (get (multiplicity p a) ((finite_mul_iff hp).1 h).1 +
+    get (multiplicity p b) ((finite_mul_iff hp).1 h).2) ∣ a * b,
+  by rw [hpoweq]; apply mul_dvd_mul; assumption,
+have hsucc : ¬p ^ ((get (multiplicity p a) ((finite_mul_iff hp).1 h).1 +
+    get (multiplicity p b) ((finite_mul_iff hp).1 h).2) + 1) ∣ a * b,
+  from λ h, not_or (is_greatest' _ (lt_succ_self _)) (is_greatest' _ (lt_succ_self _))
+    (succ_dvd_or_succ_dvd_of_succ_sum_dvd_mul hp (by convert hdiva)
+      (by convert hdivb) h),
+by rw [← enat.coe_inj, enat.coe_get, eq_some_iff];
+  exact ⟨hdiv, hsucc⟩
+
+local attribute [instance, priority 0] classical.prop_decidable
+
+protected lemma mul {p a b : α} (hp : prime p) :
+  multiplicity p (a * b) = multiplicity p a + multiplicity p b :=
+if h : finite p a ∧ finite p b then
+by rw [← enat.coe_get (finite_iff_dom.1 h.1), ← enat.coe_get (finite_iff_dom.1 h.2),
+  ← enat.coe_get (finite_iff_dom.1 (finite_mul hp h.1 h.2)),
+    ← enat.coe_add, enat.coe_inj, multiplicity.mul' hp]; refl
+else begin
+  rw [eq_top_iff_not_finite.2 (mt (finite_mul_iff hp).1 h)],
+  cases not_and_distrib.1 h with h h;
+    simp [eq_top_iff_not_finite.2 h]
+end
+
+protected lemma pow' {p a : α} (hp : prime p) (ha : finite p a) : ∀ {k : ℕ},
+  get (multiplicity p (a ^ k)) (finite_pow hp ha) = k * get (multiplicity p a) ha
+| 0     := by dsimp [_root_.pow_zero]; simp [one_right hp.2.1]; refl
+| (k+1) := by dsimp only [_root_.pow_succ];
+  erw [multiplicity.mul' hp, pow', add_mul, one_mul, add_comm]
+
+lemma pow {p a : α} (hp : prime p) : ∀ {k : ℕ},
+  multiplicity p (a ^ k) = add_monoid.smul k (multiplicity p a)
+| 0        := by simp [one_right hp.2.1]
+| (succ k) := by simp [_root_.pow_succ, succ_smul, pow, multiplicity.mul hp]
+
+end integral_domain
+
+end multiplicity
+
+section nat
+open multiplicity
+
+lemma multiplicity_eq_zero_of_coprime {p a b : ℕ} (hp : p ≠ 1)
+  (hle : multiplicity p a ≤ multiplicity p b)
```

Comment:
Is this the natural condition to put here? I would think that `p \dvd b` would be more natural...

Reply:
I guess it's natural, but not trivially equivalent, and actually this is a harder lemma to prove than the version with `p \dvd b`.
This statement is copied from `padic_val`.  I guess Rob had a use for it in this form.

Reply:
So maybe we should provide both?

Reply:
The `p \| b` version is fairly trivial to prove, and fairly niche, so I wouldn't bother with it.

Reply:
Ok, fine with me.